### ISSUE TITLE
Cannot inherit google protobuf's Message with Protocol

### ISSUE THREAD

**Issue by inq** [microsoft/pyright#1278]
Hello, Thank you for providing a great tool. But I have a trouble with using Protobuf with pyright when I am trying to inherit Message. 

I am using `pyright 1.1.95`.

1. We can use this code, as expected.
```
from typing import Generic, TypeVar
from google.protobuf.message import Message
from the_protobuf_generated_code import P

_T = TypeVar('_T', bound=Message)

class X(Generic[_T]):
    pass

x = X[T]()
```

2. But If I try to inherit the message, I cannot use the class as a bound.
```
from typing import Generic, TypeVar
from google.protobuf.message import Message
from the_protobuf_generated_code import P  # This is the protoc-generated Message (with mypy-protobuf plugin)

class People(Message):
    pass

_T = TypeVar('_T', bound=People)

class X(Generic[_T]):
    pass

x = X[P]()  # pyright regards this as an error
```
pyright throws:
```
 error: Type "P" cannot be assigned to TypeVar "_T"
    Type "P" is not compatible with bound type "People" for TypeVar "_T"
      "P" is incompatible with "People" (reportGeneralTypeIssues)
```

3. If I try to inherit Protocol also, then
```
from typing import Generic, TypeVar, Protocol
from google.protobuf.message import Message
from the_protobuf_generated_code import P  # This is the protoc-generated Message (with mypy-protobuf plugin)

class People(Message, Protocol):
    pass

_T = TypeVar('_T', bound=People)

class X(Generic[_T]):
    pass

x = X[P]()  # pyright regards this as an error
```
pyright throws the same error.

4. But If I try to use pylance (I think this uses more recent version of pyright..), the error message of (3) is changed to:
```
Type "P" cannot be assigned to TypeVar "_T"
  Type "P" is not compatible with bound type "People" for TypeVar "_T"
    "__deepcopy__" is an incompatible type
      Type "(self: P, memo: Unknown = ...) -> Unknown" cannot be assigned to type "(self: Message, memo: Unknown = ...) -> Unknown"
        Parameter 2: type "Message" cannot be assigned to type "P"
          "Message" is incompatible with "P"
    "__eq__" is an incompatible type
      Type "(self: P, other_msg: Unknown) -> Unknown" cannot be assigned to type "(self: Message, other_msg: Unknown) -> Unknown"
        Parameter 2: type "Message" cannot be assigned to type "P"
  ...Pylance(reportGeneralTypeIssues)
```

Is there any way to solve this? I thought that the code of (3) should be correct...

Thank you!

**Comment by erictraut** [comment]
I don't understand what you're trying to do with this bound TypeVar. Note that classes `P` and `People` do not derive from each other. They have a common base class `Message`, but that's all they have in common. So perhaps you meant to use a TypeVar that is bound to `Message`? What is your intent in creating a new class `People` that does not have a relationship to `P`?

You are also trying to derive a protocol class from a non-protocol class (`Message`). (And I'll point out that the `message.pyi` type stub is missing annotations for some parameters which is why you're seeing `Unknown` types.)

**Comment by inq** [comment]
Sorry for my ambiguous question. I wanted to use a TypeVar for a bound of Message with specific members, for example,

```
from typing import Generic, TypeVar, Protocol
from google.protobuf.message import Message
from the_protobuf_generated_code import P, Q  # This is the protoc-generated Message (with mypy-protobuf plugin)
# Both of P and Q has a member, named `name`

class People(Message, Protocol):
    name: str

_T = TypeVar('_T', bound=People)

class X(Generic[_T]):
    __init__(self, child: _T):
        self._child = child
        print(self._child.name)

x = X(P)  # pyright regards this as an error
y = X(Q)  # pyright regards this as an error
```

**Comment by erictraut** [comment]
You're doing something unusual (and apparently illegal) here. You're defining a protocol class (`People`) that derives from a non-protocol class (`Message`). I've gone back and reviewed [PEP 544](https://www.python.org/dev/peps/pep-0544/). I don't see anything in the PEP that disallow this, but I've never seen it done before. PEP 544 does talk about non-protocol classes that derive from protocol classes, but not vice versa. Alas, I confirmed that mypy explicitly disallows this and generates an error. For consistency, I think Pyright needs to do the same.

That means you'll need to come up with some other approach. You could define your own MessageProtocol class that duplicates the essential methods and variables that comprise a Message.

**Comment by erictraut** [comment]
I've filed the following bug in the pyright repo to track this. https://github.com/microsoft/pyright/issues/1281

**Comment by inq** [comment]
Thank you very much! Then let me do that!

**Comment by inq** [comment]
I am sorry but I have a more silly question... If I try to mimic the Message as MessageProtocol, then it will be a form of..

```
_M = TypeVar("_M", bound='MessageBound')  # message type (of self)

class MessageBound(Protocol):
    # Copied from google.protobuf.message
    DESCRIPTOR: Descriptor
    ...
    def MergeFrom(self: _M, other_msg: _M) -> None: ...
    ...
```
But If I try to use TypeVar with bound of MessageProtocol, then pyright complains:
```
Type "Account" cannot be assigned to TypeVar "_T"
  Type "P" is not compatible with bound type "MessageBound" for TypeVar "_R"
    "MergeFrom" is an incompatible type
      Type "(self: P, other_msg: P) -> None" cannot be assigned to type "(self: _M, other_msg: _M) -> None"
        Parameter 2: type "_M" cannot be assigned to type "P"
          "MessageBound" is incompatible with "P"
        Parameter 3: type "_M" cannot be assigned to type "P"
          "MessageBound" is incompatible with "P"
```
Do you have any idea about this?

**Comment by inq** [comment]
```
    def __init__(self, **kwargs) -> None: ...
```
Plus, this is incompatible also with custom constructor.... Is there any way to make a bount with any method (__init__)?

**Comment by erictraut** [comment]
When a callable type (e.g. function) is tested for type compatibility, its parameters are considered "contravariant". That means that the type of the destination parameter must be assignable to the type of the source parameter. You're defining this method in a manner in which doesn't allow for type compatibility because `_M` (which is bound to `MessageBound`) is a protocol class, and protocol classes cannot be assigned to non-protocol class.

If your goal is to simply ensure that the class has a `name` member, why are you attempting to define all of these methods in the protocol? Wouldn't it be sufficient for the protocol class to contain only a `name` member?

**Comment by erictraut** [comment]
I guess I'm still unclear about what your goal is here. If you could explain at a higher level what you're trying to do, I might be able to offer a better suggestion.

**Comment by inq** [comment]
Sorry for my ambiguous sentence. My goal is to ensure that the class has a `name` member, and with all of the Message's functionalities, for example, `ParseFromString`, `CopyFrom`, and `ListFields`, `DESCRIPTOR`, ..., etc..

**Comment by erictraut** [comment]
But why do you want to do this? What is your larger goal here?

**Comment by inq** [comment]
My larger goal is to handle some events defined by protobuf. Every events have a common member, named `name`.
There is some type of event, for example 'Signup', or 'Retire', and there are so many of different members also.
I have many of event routing class, named SignupEventRouter, and RetireEventRouter. And they are checking name and serializing the messages and send them into other server. Without typing, I can merge completely these classes, so I need only one of 'EventRouter'. It works well, but I realized that it's really hard to pass pyright test, so I wrote this article lol

**Comment by erictraut** [comment]
I guess I'm not clear why you are using generics and protocols.

If your event router is able to handle messages of any type, it could simply accept a parameter that is typed as `Message`. All of your messages are subclasses of this type, so this would all be accepted by the type checker.

If you do need to specialize the router class so it retains the message subtype, the router class could be generic, and you could use a TypeVar that is bound to `Message` (`_T = TypeVar("_T", bound=Message)`). There would be no need to create a protocol class.

**Comment by Denaun** [comment]
Hi, I bumped into this issue while trying to understand an error message that I was getting.
Since my issue is related to generics (although not protocols), I think it might help shedding some light on what @inq is seeing and possibly fixing it.

MRE:
```python
from typing import TypeVar

T = TypeVar(‘T’)


def do_nothing(x: T) -> T:
    if isinstance(x, int):
        return x
    return x
```

As far as I know there is nothing illegal in this code, and it does exactly what the signature says: takes any argument and returns something of the same type.
Still, `pyright` gives the following error message:
```
  8:16 - error: Expression of type “int” cannot be assigned to return type “T”
    Type “int” cannot be assigned to type “T” (reportGeneralTypeIssues)
```

**Comment by erictraut** [comment]
@Denaun, this behavior is "as designed". The type of `x` is being narrowed to `int` because of the `isinstance` check. And a variable of type `int` cannot be assigned to the return type `T`. Mypy gives the same error as pyright in this case.

**Comment by Denaun** [comment]
Sorry, I thought I compared with `mypy` but apparently I used slightly different code:
```python
from typing import TypeVar

T = TypeVar(‘T’, int, str)  # Only difference with the code above.

def do_nothing(x: T) -> T:
    if isinstance(x, int):
        return x
    return x
```

Here, `mypy` finds no issue, whereas `pyright` complains about both returns (once because `int` is not assignable to `T` and once for the same reason about `str`).
While `pyright` is certainly more consistent than `mypy` in this case, it feels like this behavior should not be designed like this: from a caller point of view we already have that if I call `do_nothing(1)`, its return type is `int` since that’s what `T` is resolved to. Why can’t it be the same from inside the function?

Re this not being the same issue, upon re-reading inq’s code I realize I saw generic return types where there aren’t any, sorry. I am fine with moving this to a separate issue if you think it’s worth it, otherwise we can just ignore it.

**Comment by erictraut** [comment]
I think this is a separate issue specifically related to _constrained_ type variables. The other examples in this thread are _bound_ type variables. @Denaun, could you do me a favor and open a new issue in the [pyright](https://github.com/microsoft/pyright) repo?

**Comment by erictraut** [comment]
I don't think there's anything remaining to address in this issue. Please open a new issue if there are additional bugs or questions.

**Comment by inq** [comment]
Thank you very much!
